Document updating configuration via a ConfigMap

adityasamant25 opened the pull request:
Fixes #23101

Page preview: https://deploy-preview-44540--kubernetes-io-main-staging.netlify.app/docs/tutorials/configuration/updating-configuration-via-a-configmap/ 

I have created an initial draft for "Updating configuration via a ConfigMap".
It is a step-by-step tutorial describing 2 topics:
1. Updating configuration via a ConfigMap mounted as a Volume
2. Updating environment variables of a Pod via a ConfigMap

Following this hands-on tutorial, a reader will be able to understand the difference in the behaviour for the two scenarios.

I have not added a "What's next" section, as I did not find a logical page to follow after this. Suggestions welcome.

I have linked the new page as a "What's next" link from two related pages.

This is my first attempt at documenting an entirely new page in the Kubernetes documentation.

Suggestions and feedback are welcome.

Best Regards,
Aditya.

<!--

 Hello!

 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/suggesting-improvements/

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.

-->

sftim commented:
/retitle Document updating configuration via a ConfigMap

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,322 @@
+---
+title: Updating configuration via a ConfigMap
+content_type: tutorial
```
Great. Please also set a value for `weight` (should be different from all other weights within `content/en/docs/tutorials/configuration/`.

adityasamant25 replied:
Got it. I'll add a weight.

adityasamant25 replied:
Resolved.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,322 @@
+kubectl create configmap sport --from-literal=sport=football
+```
+
+[Create a Deployment](/docs/concepts/workloads/controllers/deployment/#creating-a-deployment) manifest named `deployment.yaml` with the below specification. The ConfigMap `sport` is mounted as a Volume on the Pod.
````
Ideally, make this an example that people can apply using `kubectl apply -f <url>`. See other tutorials to see how this is done, or ask questions if you're still not sure.

adityasamant25 replied:
So I can add a new yaml to the `content/en/examples directory`, right?

sftim replied:
You are welcome to do that.

adityasamant25 replied:
I guess I need some advice about this.
I added the yaml in the examples directory and it is rendered correctly.
`examples/deployments/deployment-with-configmap-as-volume.yaml`

Now as I understand, this yaml will be uploaded to the following location once the PR is merged to main, is that right?
https://raw.githubusercontent.com/kubernetes/website/main/content/en/examples/deployments/deployment-with-configmap-as-volume.yaml 

Currently this link throws a 404 as expected.

How do I validate that everything with this example yaml is correct before the PR is merged to main?

adityasamant25 replied:
This is resolved in the latest commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,322 @@
+<!-- overview -->
+
+This page provides a step-by-step example updating configuration within a Pod via a ConfigMap and builds upon the [Configure a Pod to Use a ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) task. 
+
```
The example uses NGINX; I would mention NGINX here and also be clear that we chose that just as an example.

adityasamant25 replied:
Resolved.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,322 @@
+    spec:
+      containers:
+      - name: nginx
+        image: nginx:1.14.2
```
Consider using https://hub.docker.com/r/nginxinc/nginx-unprivileged, which does not require running as `root`.

adityasamant25 replied:
Clear, thanks for sending a reference to this image. I will use the nginx-unprivileged as an example.

adityasamant25 replied:
Resolved.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,322 @@
+          - name: config-volume
+            mountPath: /etc/config
+      volumes:
+        - name: config-volume
+          configMap:
+            name: sport
```
How does this ConfigMap affect the behavior of the Pod?

adityasamant25 replied:
I will extend this to actually use the value in a log to stdout.

adityasamant25 replied:
I modified the tutorial to use the values from the ConfigMap in logs printed to stdout. This should be sufficient as I want to keep the application used in the tutorial as simple as possible.

sftim reviewed (commented):
Overall:

- the ConfigMap must change the behavior of the Pod. Without that detail, it is not doing configuration.
- we should explain when and why you might need to roll out the change by updating the associated Deployment

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,322 @@
+
+Check the pod.
+```shell
+kubectl get pod <pod-name>
````
How does the reader know what the name of the Pod is?

adityasamant25 replied:
Acknowledged. I will improve this.

adityasamant25 replied:
Resolved.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,322 @@
+  labels:
+    app: nginx
+spec:
+  replicas: 1
```
BTW, it's more realistic to have at least 2 replicas.

adityasamant25 replied:
Acknowledged.

adityasamant25 replied:
Resolved, increased replicas to 3.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,322 @@
+
+Edit the deployment `configmap-test` to add an annotation to the Pod.
+```shell
+kubectl edit deployment configmap-test
````
I'd use `kubectl rollout`.

adityasamant25 replied:
Agreed, this is a much better option. Thanks for pointing it out. I've resolved it.

adityasamant25 commented:
@sftim Many thanks for the detailed review, I've implemented the suggested changes. Would love your feedback on this and I'm available to implement any further improvements that may be needed here.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,256 @@
+---
+title: Updating configuration via a ConfigMap
```
(tiny nit) page titles should be in title case, according to the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/)

This feedback shouldn't block a merge.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,256 @@
+
+<!-- overview -->
+This page provides a step-by-step example of updating configuration within a Pod via a ConfigMap and builds upon the [Configure a Pod to Use a ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) task. At the end of this tutorial, you will understand how to change the configuration for a running application. 
+This tutorial uses the NGINX image as an example.
```
Why NGINX? We could use something else, because we're not using that image for its usual purpose of being a webserver.

You could change the image to Alpine Linux or Debian and I think it would still work OK. Is that right?

---

Extra credit: add a final step that runs _two_ containers, one writing a file in HTML format to an `emptyDir`, and the other serving the HTML via HTTP. Change the HTML based on a ConfigMap.

adityasamant25 replied:
I agree. In the latest commit, I have modified the examples to use the ```alpine``` image.

adityasamant25 replied:
I'll work on the final step of a Pod with two containers, rendering HTML on the basis of a ConfigMap.

adityasamant25 replied:
In the latest commit, I have added a third section that demonstrates the scenario with two containers.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
Sports are fine, but for vehicle brands, let's avoid referring to real companies.

Maybe swap `car` for `fruit`?

adityasamant25 replied:
I agree. I have changed the example to use ```fruit``` and it is available in the latest commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,256 @@
+data:
+  sport: cricket
+kind: ConfigMap
+metadata:
+  creationTimestamp: "2023-12-27T08:24:51Z"
+  name: sport
+  namespace: default
+  resourceVersion: "1467267"
+  uid: d2d51452-8a70-40d8-b61a-0a2e92f3039b
```
```suggestion
# You can leave the existing metadata as they are.
# The values you'll see won't exactly match these.
metadata:
  creationTimestamp: "2023-12-27T08:24:51Z"
  name: sport
  namespace: default
  resourceVersion: "1467267"
  uid: d2d51452-8a70-40d8-b61a-0a2e92f3039b
```

adityasamant25 replied:
Thanks for the suggestion.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,256 @@
+```
+After few seconds, you should see the log output change as follows:
+```
+FROM:
+My preferred sport is football
+TO:
+My preferred sport is cricket
````
You could try a `diff` format block. The `FROM:` and `TO:` aren't literal console output; don't make it look like that is what the reader would see.

Eg
```console
My preferred sport is football
My preferred sport is football
My preferred sport is football
My preferred sport is cricket
```

Extra credit: include a timestamp in the log lines.

adityasamant25 replied:
Done. It is implemented in the latest commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,256 @@
+My preferred sport is cricket
+```
+
+When a ConfigMap mounted as a Volume is updated, the projected content is seamlessly updated too.
````
```suggestion
When you have a ConfigMap that is mapped into a running Pod using either a `configMap` volume or a `projected` volume, and you update that ConfigMap, the running Pod sees the update almost immediately.
```

adityasamant25 replied:
Thanks for the suggestion.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,256 @@
+
+This demonstrates the scenario of updating environment variables in a Pod that are derived from a ConfigMap. Changes to the ConfigMap values are applied to the Pod during the subsequent rollout.  
+
+**Summary:**  
```
```suggestion
## Summary

```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,256 @@
+kubectl edit configmap sport
+```
+
+In the yaml editor, change the value of key `sport` from `football` to `cricket` and save.
````
```sugggestion
In the editor that appears, change the value of key `sport` from `football` to `cricket. Save your changes.
The kubectl tool updates the ConfigMap accordingly (if you see an error, try again).

Here's an example of how that manifest could look after you edit it:
```

adityasamant25 replied:
Thanks for this suggestion, it is clearly better. I'll add this in my next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,256 @@
+
+Tail the logs of the Pod.
+```shell
+kubectl logs -f <pod-name>
````
Does this work?
```suggestion
kubectl logs -f deployments/configmap-volume
```

adityasamant25 replied:
Yes, it works. I have implemented this in the latest commit.

sftim reviewed (commented):
Thanks. Some more feedback.

adityasamant25 commented:
@sftim, Thank you for the review. I have implemented all the feedback. The tutorial looks much better now. 
Can you please take a look?

adityasamant25 commented:
@sftim @kbhawkey @eparis,
Can you please take a look at the latest draft of this new tutorial on "Updating configuration via a ConfigMap"?

I would really love for this tutorial to be published to the official documentation and I'm available for any improvements to be made.

adityasamant25 commented:
/assign @sftim 

Hi @sftim,
The PR is ready for another review after implementing the previous review comments.
PTAL.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+configmap-volume-6b976dfdcf-tbc6r   1/1     Running   0          72s
+```
+
+The data in the ConfigMap is mounted to the Pods as a Volume. The Pod references this data to generate an output that is printed to stdout. Check this by viewing the logs of the Pods.
````
```suggestion
On each node where one of these Pods is running, the kubelet fetches the data for that ConfigMap
and translates it to files in a local volume. The kubelet then mounts that volume into the container,
as specified in the Pod template. The code running in that container loads the information from the
file and uses it printed a report to stdout. You can check this report by viewing the logs for one of the
Pods in that Deployment:
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+
+The data in the ConfigMap is mounted to the Pods as a Volume. The Pod references this data to generate an output that is printed to stdout. Check this by viewing the logs of the Pods.
+```shell
+kubectl logs deployments/configmap-volume
````
```suggestion
# Pick one Pod that belongs to the Deployment, and view its logs
kubectl logs deployments/configmap-volume
```

adityasamant25 replied:
This should not be necessary as `kubectl logs deployments/configmap-volume` picks one of the Pod at random and displays its logs. This is seen in the output of the command, similar to `Found 3 pods, using pod/configmap-volume-76d9c5678f-x5rgj`

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+Thu Jan  4 14:07:26 UTC 2024 My preferred sport is football
+```
+
+Edit the ConfigMap.
````
```suggestion
Edit the ConfigMap:
```

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+Thu Jan  4 14:12:16 UTC 2024 My preferred sport is cricket
+```
+
+When you have a ConfigMap that is mapped into a running Pod using either a `configMap` volume or a `projected` volume, and you update that ConfigMap, the running Pod sees the update almost immediately.
````
```suggestion
When you have a ConfigMap that is mapped into a running Pod using either a `configMap` volume
or a `projected` volume, and you update that ConfigMap, the running Pod sees the update almost
immediately. However, your application only sees the change if it is written to either poll for changes,
or watch for file updates. An application that loads its configuration once at startup will not notice
a change.
```

adityasamant25 replied:
Ack, very nice suggestion. It will be included in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+kubectl logs deployment/configmap-env-var
+```
+
+You will notice that the output remains unchanged:
````
```suggestion
Notice that the output remains **unchanged**, even though you edited the ConfigMap:
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+```
+
+{{< note >}}
+Although the value of the key inside the ConfigMap has changed, the environment variable in the Pod still possesses the earlier value. This is because environment variables in a Pod are not updated until the Pod is rolled-over.
````
```suggestion
Although the value of the key inside the ConfigMap has changed, the environment variable in the Pod still
shows the earlier value. This is because environment variables for a process running inside a Pod are
**not** updated when the source data changes; if you wanted to force an update, you would need to
have Kubernetes replace your existing Pods. The new Pods would then run with the updated information.
```

adityasamant25 replied:
Ack, will be included in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,381 @@
+Although the value of the key inside the ConfigMap has changed, the environment variable in the Pod still possesses the earlier value. This is because environment variables in a Pod are not updated until the Pod is rolled-over.
+{{< /note >}}
+
+Let's perform a rollout for the Deployment that will cause all the Pods to restart.
```
```suggestion
You can trigger that replacement. Perform a rollout for the Deployment, using
[`kubectl rollout`](/docs/reference/kubectl/generated/kubectl_rollout/):
```

adityasamant25 replied:
Ack, will be included in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+kubectl rollout restart deployment configmap-env-var
+```
+
+Check the deployment.
````
```suggestion
Next, check the Deployment:
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+configmap-env-var   3/3     3            3           12m
+```
+
+Check the Pods.
````
```suggestion
Check the Pods:
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+
+Let's perform a rollout for the Deployment that will cause all the Pods to restart.
+```shell
+kubectl rollout restart deployment configmap-env-var
````
```suggestion
# Trigger the rollout
kubectl rollout restart deployment configmap-env-var

# Wait for the rollout to complete
kubectl rollout status deployment configmap-env-var --watch=true
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+kubectl get pods --selector=app=configmap-env-var
+```
+
+The older Pods will terminate and new ones will be created. You should finally see an output similar to:
````
```suggestion
The rollout caused Kubernetes to make a new
{{< glossary_tooltip term_id="replica-set" text="ReplicaSet" >}} for the Deployment; that means that the
existing Pods were terminated, and new ones have be created. You should finally see an output similar to:
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+
+View the logs of the Deployment.
+```shell
+kubectl logs deployment/configmap-env-var
````
```suggestion
# Pick one Pod that belongs to the Deployment, and view its logs
kubectl logs deployment/configmap-env-var
```

adityasamant25 replied:
This should not be necessary as kubectl logs deployments/configmap-volume picks one of the Pod at random and displays its logs.

sftim replied:
We're explaining what the command does, here.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+configmap-env-var-6d94d89bf5-d5vx8   1/1     Running       0          11s
+```
+
+View the logs of the Deployment.
````
```suggestion
View the logs for a Pod in this Deployment:
```

adityasamant25 replied:
Ack

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+Thu Jan  4 16:30:55 UTC 2024 The basket is full of mangoes
+```
+
+This demonstrates the scenario of updating environment variables in a Pod that are derived from a ConfigMap. Changes to the ConfigMap values are applied to the Pod during the subsequent rollout.  
````
```suggestion
This demonstrates the scenario of updating environment variables in a Pod that are derived
from a ConfigMap. Changes to the ConfigMap values are applied to the Pod during the subsequent
rollout. If Pods get created for other reason, such as scaling up the Deployment, then the new Pods
also use the latest configuration values; if you don't trigger a rollout, then you might find that your
app is running with a mix of old and new environment variable values.
```

adityasamant25 replied:
Great addition. Thanks. Ack.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-two-containers.yaml
+```
+
+Check the deployment.
````
```suggestion
Check the Deployment:
```

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+Below is an example of a Deployment having Pods with two containers. The two containers share an ```emptyDir``` Volume that they use to communicate. The first container runs a ```nginx``` server. The mount path for the shared Volume is ```/usr/share/nginx/html```. The second container is based on the ```alpine``` image, and has a mount path of ```/pod-data```. The ```alpine``` container writes a  ...
+{{% code_sample file="deployments/deployment-with-configmap-two-containers.yaml" %}}
+
+Create the deployment.
````
```suggestion
Create the Deployment:
```

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+
+```
+
+Check the pods.
````
```suggestion
Check the pods:
```

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+kubectl get pods --selector=app=configmap-two-containers
+```
+You should see an output similar to:
+```shell
````
This output isn't a shell script; drop the `shell` part.

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
`````diff
@@ -0,0 +1,381 @@
+configmap-two-containers-565fb6d4f4-mzsmf   2/2     Running   0          20s
+````
+
+Expose the deployment via a Service.
`````
```suggestion
Expose the Deployment (the `kubectl` tool creates a
{{<glossary_tooltip text="Service" term_id="service">}}  for you):
```

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+
+Use ```kubectl``` to forward the port:
+```shell
+kubectl port-forward service/configmap-service 8080:8080 &
````
```suggestion
kubectl port-forward service/configmap-service 8080:8080 & # this stays running in the background
```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+
+Access the service.
+```shell
+curl localhost:8080
````
```suggestion
curl http://localhost:8080
```

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+```
+
+You should see an output similar to:
+```shell
````
This output isn't a shell script; drop the `shell` part.

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+
+Loop over the service URL for few seconds. 
+```shell
+while true; do curl localhost:8080; sleep 10; done
````
How about:
```suggestion
# Cancel this when you're happy with it (Ctrl-C)
while true; do curl --connect-timeout 7.5 http://localhost:8080; sleep 10; done
```

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,381 @@
+
+## {{% heading "prerequisites" %}}
+{{< include "task-tutorial-prereqs.md" >}}
+
```
Please mention that you need `curl` (it's really popular, but even so, let's be clear about what people need to have).

adityasamant25 replied:
Ack. I'm adding it as a new page under the `includes` directory to make it reusable. Any tutorial that needs `cURL` as a prerequisite can include this page in their tutorial to save time and have a standard pattern across tutorials.

sftim reviewed (commented):
Thanks. Here's a further review round.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,381 @@
+Below is an example of a Deployment manifest with the ConfigMap `sport` mounted as a Volume on the Pod.
+{{% code_sample file="deployments/deployment-with-configmap-as-volume.yaml" %}}
+
+Create the deployment.
```
```suggestion
Create the Deployment:
```

adityasamant25 replied:
Ack, will be included in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-as-volume.yaml
+```
+
+Check the deployment.
````
```suggestion
Check the Deployment:
```

adityasamant25 replied:
Ack, will be included in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+configmap/sport edited
+```
+
+Tail the logs of the Deployment.
````
```suggestion
Tail (follow the latest entries in) the logs of one of the pods that belongs to this Deployment:
```

adityasamant25 replied:
Ack, will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,381 @@
+Below is an example of a Deployment manifest with an environment variable configured via the ConfigMap `fruits`.
+{{% code_sample file="deployments/deployment-with-configmap-as-envvar.yaml" %}}
+
+Create the deployment.
```
(nit)
```suggestion
Create the Deployment:
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+kubectl apply -f https://k8s.io/examples/deployments/deployment-with-configmap-as-envvar.yaml
+```
+
+Check the deployment.
````
```suggestion
Check the Deployment:
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+configmap-env-var   3/3     3            3           7s
+```
+
+Check the pods.
````
```suggestion
Check the pods for this Deployment (matching by
{{< glossary_tooltip text="selector" term_id="selector" >}}):
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+kubectl logs deployment/configmap-env-var
+```
+
+You should see an output similar to the below:
````
More idiomatic as English:
```suggestion
You should see an output similar to:
```

adityasamant25 replied:
Ack, will be part of the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+Thu Jan  4 16:07:26 UTC 2024 The basket is full of apples
+```
+
+Edit the ConfigMap.
````
```suggestion
Edit the ConfigMap:
```

adityasamant25 replied:
Ack, will be part of the next commit.

adityasamant25 commented:
@sftim Thanks for the comprehensive review, I really appreciate it. I'll be working on the comments.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,381 @@
+---
+title: Updating Configuration via a Configmap
+content_type: tutorial
+weight: 20
```
Ack, will be handled in the next commit.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+configmap-volume   3/3     3            3           19s
+```
+
+Check the pods.
````
Ack, will be handled in the next commit.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+```
+
+You should see the output change as follows:
+```shell
````
Ack, will be handled in the next commit.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+Use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):
+
+```shell
+kubectl create configmap colour --from-literal=colour=red
````
Ack, will be handled in the next commit.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+kubectl create configmap colour --from-literal=colour=red
+```
+
+Below is an example of a Deployment having Pods with two containers. The two containers share an ```emptyDir``` Volume that they use to communicate. The first container runs a ```nginx``` server. The mount path for the shared Volume is ```/usr/share/nginx/html```. The second container is based on the ```alpine``` image, and has a mount path of ```/pod-data```. The ```alpine``` container writes a  ...
````
Ack, will be handled in the next commit.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+Fri Jan  5 08:08:22 UTC 2024 My preferred colour is red
+```
+
+Edit the ConfigMap.
````
Ack, will be handled in the next commit.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+Fri Jan  5 08:14:43 UTC 2024 My preferred colour is blue
+Fri Jan  5 08:15:00 UTC 2024 My preferred colour is blue
+```
+
````
I will work on adding this, however it may need some time. Once its ready, we can add it to the current PR if the PR is still open. Else, I can create a new PR for it later.

I believe this should not be a blocker to merge the current PR.

adityasamant25 replied:
Added a 4th scenario with the Sidecar Container.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,381 @@
+```shell
+kubectl create configmap sport --from-literal=sport=football
+```
+Below is an example of a Deployment manifest with the ConfigMap `sport` mounted as a Volume on the Pod.
````
Ack. Thanks for this suggestion to use tooltips. I learnt something new. :)

adityasamant25 commented on content/en/examples/deployments/deployment-with-configmap-two-containers.yaml:
```diff
@@ -0,0 +1,40 @@
+  replicas: 3
+  selector:
+    matchLabels:
+      app: configmap-two-containers
```
Ack

adityasamant25 commented on content/en/examples/deployments/deployment-with-configmap-as-volume.yaml:
```diff
@@ -0,0 +1,33 @@
+  replicas: 3
+  selector:
+    matchLabels:
+      app: configmap-volume
```
Ack

adityasamant25 commented on content/en/examples/deployments/deployment-with-configmap-as-envvar.yaml:
```diff
@@ -0,0 +1,32 @@
+  replicas: 3
+  selector:
+    matchLabels:
+      app: configmap-env-var
```
Ack.

adityasamant25 commented:
@sftim I have implemented all the suggestions and pushed a new commit.

The preview page can be seen [here](https://deploy-preview-44540--kubernetes-io-main-staging.netlify.app/docs/tutorials/configuration/updating-configuration-via-a-configmap/).

The only comment pending is the suggestion to add a fourth example depicting the scenario for the sidecar containers. This may take some time, however in my opinion, it should not block the merge of the PR.

After the corrections, I have implemented all the steps of the tutorial on my machine and verified that all the examples work well.

Thanks again for the great review, it makes the page look much better.

May I request you to take another look and approve?

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,403 @@
+## {{% heading "prerequisites" %}}
+{{< include "task-tutorial-prereqs.md" >}}
+
+{{< include "task-tutorial-curl-prereq.md" >}}
```
Thanks, I have implemented this.

adityasamant25 commented on content/en/includes/task-tutorial-curl-prereq.md:
```diff
@@ -0,0 +1 @@
+You need to have the `cURL` command-line tool for invoking HTTP requests from the terminal or command prompt. If you do not have `cURL` installed, you can install it by following the instructions specified [here](/docs/tasks/tools/curl-prerequisite).
```
As per the other comment, I have removed this new file.

adityasamant25 commented:
@sftim, Thanks for the latest review. The changes are done and available for preview [here](https://deploy-preview-44540--kubernetes-io-main-staging.netlify.app/docs/tutorials/configuration/updating-configuration-via-a-configmap/).

PTAL.

adityasamant25 commented:
@sftim, I have added a 4th example depicting the use of the Sidecar Container.

sftim commented:
Relevant to https://github.com/kubernetes/website/issues/1911

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,514 @@
+
+Tail (follow the latest entries in) the logs of one of the pods that belongs to this Deployment:
+```shell
+kubectl logs -f deployments/configmap-volume
````
I recommend
```suggestion
kubectl logs deployments/configmap-volume --follow
```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,514 @@
+
+Tail the logs of the Deployment and observe the output for few seconds:
+```shell
+kubectl logs deployment/configmap-env-var
````
I recommend
```suggestion
# As the text explains, the output does NOT change
kubectl logs deployments/configmap-env-var --follow
```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,514 @@
+# Wait for the rollout to complete
+kubectl rollout status deployment configmap-env-var --watch=true
+```
+
````
There's a problem; `kubectl rollout` does not actually wait for the old Pods to disappear. The old Pods will still be around, for a short time, albeit `Terminating`.

sftim replied:
This works, but what a lot to type:
```shell
kubectl rollout restart deployment configmap-env-var && \
  kubectl rollout status deployment configmap-env-var --watch=true && \
  while kubectl get pods --selector=app.kubernetes.io/name=configmap-env-var \
  --template '{{range .items}}{{ if .metadata.deletionTimestamp }}pods/{{.metadata.name}}{{"\n"}}{{end}}{{end}}' |
  grep -q . ; do sleep 1.5; done && echo Rollout complete 1>&2
```

(I wonder if `kubectl rollout` should support waiting for the selected Pods to actually terminate, or if the Deployment and ReplicaSet controllers should somehow track the Pods that they once owned right through to deletion)

adityasamant25 replied:
Thanks for the alternate command, but I agree it is too much to type for this tutorial and we would not want to confuse the readers.

I will handle it by modifying the text to clearly imply that the readers should wait till the older Pods terminate completely before proceeding with the tutorial.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,514 @@
+configmap-env-var-6d94d89bf5-d5vx8   1/1     Running       0          11s
+```
+
+View the logs for a Pod in this Deployment:
````
There's a problem: we haven't waited for the previous Pods to terminate.

This works, but it's a bit long to type:
```shell
kubectl logs --follow \
  "$( kubectl get pods --selector=app.kubernetes.io/name=configmap-env-var \
  --template '{{range .items}}{{ if not .metadata.deletionTimestamp }}pods/{{.metadata.name}}{{"\n"}}{{end}}{{end}}' |
  head -n 1 )"
```

adityasamant25 replied:
Thanks for the alternate command, but I agree it is too much to type for this tutorial and we would not want to confuse the readers.

I will handle it by modifying the text to clearly imply that the readers should wait till the older Pods terminate completely before proceeding with the tutorial.

sftim replied:
I've marked this unresolved until a fix is confirmed.

adityasamant25 replied:
I've addressed this in the latest commit.

sftim reviewed (commented):
Thanks. This review may raise more questions than it answers.

dipesh-rawat commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
@adityasamant25 Great job on creating this new page! I noticed that there's a request in the [issue](https://github.com/kubernetes/website/issues/23101#issue-677612881)  to include instructions on how to make this work with immutable ConfigMaps. I don't see that content in the PR. Is it something I missed or it's not included in the scope of this update?

adityasamant25 replied:
@dipesh-rawat Thanks for noting it, the scenario for immutable ConfigMaps had slipped my mind. I'll work on including it.

adityasamant25 replied:
Related issue #42359

adityasamant25 commented:
The latest commit adds a scenario for Immutable ConfigMaps.
Issue #42359 is related, I've read the discussion in the issue and tried to add the learnings to the scenario.

Updating configuration for Immutable ConfigMaps is tricky.

I've added warnings that this should be avoided as far as possible.

Note, after recreating a Immutable ConfigMap, a simple `kubectl rollout restart` is insufficient to refresh the ConfigMap content. The only thing that works is deleting the Deployment and recreating it.

Another option would be to create a new ConfigMap (rename) and reference this ConfigMap instead, however I have not documented this as I wanted to focus on how to achieve the configuration change without creating an additional ConfigMap.

Comments and feedback (especially on the Immutable ConfigMap scenario) are welcome.

sftim commented on content/en/examples/configmap/immutable-configmap.yaml:
```diff
@@ -0,0 +1,7 @@
+apiVersion: v1
+data:
+  season: "summer"
+kind: ConfigMap
+immutable: true
+metadata:
+  name: season
```
I think this isn't an ideal example of immutable data. Can we think of a setting that a human coder would reasonably think “yes, I can see why that wouldn't change very often”?

How about:
```yaml
data:
  cloud_provider_region: antarctica-1
```

or
```yaml
data:
  operating_system: Linux
```

adityasamant25 replied:
Thanks. I have implemented the example with the company names. It is available in the latest commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,723 @@
+Sat Feb 17 13:13:35 UTC 2024 My preferred color is green
+```
+
+## Update configuration via a Immutable ConfigMap mounted as a Volume
````
```suggestion
## Update configuration via an immutable ConfigMap that is mounted as a Volume {#rollout-configmap-immutable-volume}
```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,723 @@
+
+## Update configuration via a Immutable ConfigMap mounted as a Volume
+
+{{< warning >}}
+Immutable ConfigMaps are especially used for configuration which is constant and is not expected to change over time.   
+In a production use-case, it is advisable to not change the contents of Immutable ConfigMaps unless absolutely necessary.
+{{< /warning >}}
```
Warning is too strong
```suggestion
{{< note >}}
Immutable ConfigMaps are especially used for configuration that is constant and is **not**
expected to change over time. Marking a ConfigMap allows a performance improvement where the
kubelet does not watch for changes.

If you do need to make a change, you should plan to either:
- change the name of the ConfigMap, and switch to running Pods that reference the new name
- replace all the nodes in your cluster that have previously run a Pod that used the old value
- restart the kubelet on any node where the kubelet previously loaded the old ConfigMap
{{< /note >}}
```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,723 @@
+```
+
+Once a ConfigMap is marked as immutable, it is not possible to revert this change nor to mutate the contents of the data or the binaryData field.  
+In order to modify the contents, the only option is to delete and recreate the ConfigMap.
+
+Delete and recreate the ConfigMap `season` by using the temporary manifest as follows:
+```shell
+# Use the appropriate path to the temporary manifest
+kubectl replace --force -f /var/folders/nn/bb6y1j8d69n9lhqppjlrfthm0000gn/T/kubectl-edit-139284375.yaml
+```
````
This isn't always enough, because the kubelet doesn't watch for changes to immutable ConfigMaps.

Instead, I would put a date in the ConfigMap name, and teach switching to referencing a new different ConfigMap.

For example:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: "company-name-20150801"
data:
  company_name: "ACME, Inc." # existing fictional company name
immutable: true
```

and then switch to something like:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: "company-name-20240312"
data:
  company_name: "Fiktivesunternehmen GmbH" # new fictional company name
immutable: true
```

adityasamant25 replied:
This is implemented in the latest commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,723 @@
+Tue Mar  5 12:02:37 UTC 2024 My preferred season is summer
+```
+
+{{< caution >}}
+Although the ConfigMap has been recreated, the existing Pods maintain a mount point to the deleted ConfigMap. To force an update, you would need to clean up every existing reference to the old ConfigMap. In this scenario, it equates to deleting the Deployment and recreating it. Exercise caution before executing these steps in a production environment.
+
+A simple restart of the existing Deployment using the `kubectl rollout restart` command causes the Pods to be recreated, but is insufficient to refresh the contents of the ConfigMap.
+{{< /caution >}}
````
```suggestion
{{< note >}}
A simple restart of the existing Deployment using the `kubectl rollout restart` command causes the Pods
to be recreated, but is insufficient to refresh the contents of the ConfigMap.
{{< /note >}}

Although the ConfigMap has been recreated, any existing Pods maintain a mount point to the volume
representing the deleted ConfigMap. To force an update, you would need to clean up every existing
reference to the old ConfigMap. In this scenario, it equates to deleting the Deployment and recreating it.
You should be cautious about executing similar steps in a production environment.
```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,723 @@
+
+Changes to a ConfigMap that configures environment variables for a Pod are available after the subsequent rollout for the Pod.
+
+Once a ConfigMap is marked as immutable, it is not possible to revert this change nor to mutate the contents of the data or the binaryData field. You can only delete and recreate the ConfigMap. Because existing resources maintain a mount point to the deleted ConfigMap, it is necessary to ensure that every existing reference to the old ConfigMap (not just Pods, any reference) has been cleaned up.
```
```suggestion
Once a ConfigMap is marked as immutable, it is not possible to revert this change
(you cannot make an immutable ConfigMap mutable), and you also cannot make any change to
the contents of the `data` or the `binaryData` field. You can delete and recreate the ConfigMap,
or you can make a new different ConfigMap. When you delete a ConfigMap, running containers
and their Pods maintain a mount point to any volume that referenced that existing ConfigMap.
```

sftim commented on content/en/examples/deployments/deployment-with-immutable-configmap-as-volume.yaml:
```diff
@@ -0,0 +1,33 @@
+          command:
+            - /bin/sh
+            - -c
+            - while true; do echo "$(date) My preferred season is $(cat /etc/config/season)";
```
```suggestion
            - while true; do echo "$(date) The current season is $(cat /etc/config/season)";
```
or, if following other advice:
```shell
            - while true; do echo "$(date) The name of the company is $(cat /etc/config/company_name)";
```

adityasamant25 replied:
I have aligned this with the example of the company names.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,723 @@
+
+<!-- lessoncontent -->
+
+## Update configuration via a ConfigMap mounted as a Volume
```
```suggestion
## Update configuration via a ConfigMap mounted as a Volume {#rollout-configmap-volume}
```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,723 @@
+app is running with a mix of old and new environment variable values.
+
+
+## Update configuration via a ConfigMap in a multi-container Pod
```
```suggestion
## Update configuration via a ConfigMap in a multi-container Pod {#rollout-configmap-multiple-containers}
```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,723 @@
+Fri Jan  5 08:15:00 UTC 2024 My preferred color is blue
+```
+
+## Update configuration via a ConfigMap in a Pod possessing a Sidecar Container
````
```suggestion
## Update configuration via a ConfigMap in a Pod possessing a sidecar container {#rollout-configmap-sidecar}
```

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,723 @@
+
+## Update configuration via a ConfigMap in a Pod possessing a Sidecar Container
+
+The above scenario can be replicated by using a [Sidecar Container](/docs/concepts/workloads/pods/sidecar-containers/) as a helper container to write the HTML file. As a Sidecar Container is conceptually an Init Container, it is guaranteed to start before the main webserver container. This ensures that the HTML file is always available when the webserver is ready to serve it. Please see [Enabling ...
+
+If you are continuing from the previous scenario, you can reuse the ConfigMap named `color` for this scenario. If you are executing this scenario independently, use the `kubectl create configmap` command to create a ConfigMap from [literal values](/docs/tasks/configure-pod-container/configure-pod-configmap/#create-configmaps-from-literal-values):
```
Please wrap these lines.

adityasamant25 replied:
This is addressed in the latest commit.

adityasamant25 commented:
@sftim Many thanks for the latest review, I highly appreciate and value your suggestions. I have implemented all the changes in the latest commit. Kindly take a look.

tengqm commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,724 @@
+Thu Jan  4 14:12:16 UTC 2024 My preferred sport is cricket
+```
+
+When you have a ConfigMap that is mapped into a running Pod using either a `configMap` volume or a `projected` volume, and you update that ConfigMap, the running Pod sees the update almost immediately. However, your application only sees the change if it is written to either poll for changes, or watch for file updates. An application that loads its configuration once at startup will not notice a  ...
````
Please wrap these long lines manually.

adityasamant25 replied:
Done.

tengqm commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,724 @@
+Thu Jan  4 16:13:26 UTC 2024 The basket is full of apples
+```
+
+{{< note >}}
+Although the value of the key inside the ConfigMap has changed, the environment variable in the Pod still shows the earlier value. This is because environment variables for a process running inside a Pod are **not** updated when the source data changes; if you wanted to force an update, you would need to have Kubernetes replace your existing Pods. The new Pods would then run with the updated info ...
+{{< /note >}}
````
Since this subsection only differs from the previous one on how the changes are effected, I'd suggest we focus on the differences. Even for a tutorial, we don't need to enumerate each and every steps.

adityasamant25 replied:
I believe it will help the reader to copy/paste the commands for ease of learning. So I propose to keep the steps..

adityasamant25 replied:
Thanks, I've made the necessary changes.

tengqm replied:
If you want to keep this subsection rather than drop it completely, you may want to focus on the "meaningful" steps, which is that users have to recreate their pods. There are many ways to recreate them. The simplest one is to delete and recreate.
I mean we don't need the lines on creating and verifying the ConfigMap or the Deployment.

adityasamant25 replied:
The deployment manifest in the section for ConfigMap as environment variable is different to the previous section where the ConfigMap is mounted as a Volume. So I think it is necessary to provide the exact manifest on how to create the deployment, so that the reader does not make a mistake here.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,724 @@
+---
+reviewers:
+  - sftim
```
```suggestion
```
I don't have any special ConfigMap skills, so I wouldn't list me here.

adityasamant25 replied:
I've handled this in the latest commit.

adityasamant25 commented on content/en/docs/tutorials/configuration/configure-redis-using-configmap.md:
```diff
@@ -4,6 +4,7 @@ reviewers:
 - pmorie
 title: Configuring Redis using a ConfigMap
 content_type: tutorial
+weight: 30
```
I have confirmed this.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,724 @@
+---
+reviewers:
+  - sftim
+title: Updating Configuration via a Configmap
```
Thanks, I've addressed this in the latest commit.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,724 @@
+---
+
+<!-- overview -->
+This page provides a step-by-step example of updating configuration within a Pod via a ConfigMap and builds upon the [Configure a Pod to Use a ConfigMap](/docs/tasks/configure-pod-container/configure-pod-configmap/) task. At the end of this tutorial, you will understand how to change the configuration for a running application. 
```
Done.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,724 @@
+```
+NAME               READY   UP-TO-DATE   AVAILABLE   AGE
+configmap-volume   3/3     3            3           19s
+```
````
Thanks, I've removed the deployment check from all relevant places. At the place where a kubectl rollout restart is performed, I've kept the deployment check as its important to verify the rollout was completed successfully.

adityasamant25 commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,724 @@
+configmap-volume-6b976dfdcf-qxvbm   1/1     Running   0          72s
+configmap-volume-6b976dfdcf-skpvm   1/1     Running   0          72s
+configmap-volume-6b976dfdcf-tbc6r   1/1     Running   0          72s
+```
````
It is important to verify that the Pods are in a ready state before proceeding with the tutorial. So I propose to keep this step.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,673 @@
+```
+
+In the editor that appears, perform the following actions:
+- change the name of the ConfigMap from `company-name-20150801` to `company-name-20240312`
````
does this really work?

adityasamant25 replied:
All instructions have been tested by me and they work successfully.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,675 @@
+error: At least one of apiVersion, kind and name was changed
+```
+
+Create a new ConfigMap by referencing the temporary manifest.
+```shell
+kubectl apply -f /var/folders/nn/bb6y1j8d69n9lhqppjlrfthm0000gn/T/kubectl-edit-1440698061.yaml
+```
+
+You should see an output similar to:
+```
+configmap/company-name-20240312 created
+```
````
Should they delete the old ConfigMap?

adityasamant25 replied:
It is not necessary to delete the old ConfigMap.

sftim commented:
/lgtm

Adding this new page makes the site better than if we didn't have a tutorial for it.

tengqm commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,675 @@
+configmap-volume-6b976dfdcf-tbc6r   1/1     Running   0          72s
+```
+
+On each node where one of these Pods is running, the kubelet fetches the data for that ConfigMap and translates it to files in a local volume. The kubelet then mounts that volume into the container, as specified in the Pod template. The code running in that container loads the information from the file and uses it to print a report to stdout. You can check this report by viewing the logs for one  ...
````
Please do wrap long lines like this.

adityasamant25 replied:
Ack. This will be handled in the next commit.

tengqm commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,675 @@
+
+
+{{< note >}}
+The total delay from the moment when the ConfigMap is updated to the moment when new keys are projected to the Pod can be as long as kubelet sync period.  
```
Please wrap the long lines.

adityasamant25 replied:
Ack. This will be handled in the next commit.

tengqm commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,675 @@
+Also check [Mounted ConfigMaps are updated automatically](/docs/tasks/configure-pod-container/configure-pod-configmap/#mounted-configmaps-are-updated-automatically).
+{{< /note >}}
+
+## Update environment variables of a Pod via a ConfigMap {#rollout-configmap-env}
```
I don't think we need this subsection. In a tutorial, we show users how to get something done, step by step. It is weird to read something like "hey, follow me, I'm showing you that something doesn't work."

It is for sure that we need a callout, maybe something like a warning, to tell users that this update cannot apply to environment variables. We can briefly explain why it doesn't work. However, having a dedicated subsection here showing that something doesn't work sounds an overkill.

adityasamant25 replied:
Once a reader follows these steps, the difference between ConfigMap as Volume and as env variable will be crystal clear. I think the section has merit, and reducing it to a callout may be insufficient to demonstrate the difference.

I propose to keep the section, in any case it will not do any harm :)

tengqm commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
`````diff
@@ -0,0 +1,675 @@
+configmap-two-containers-565fb6d4f4-mzsmf   2/2     Running   0          20s
+````
+
+Expose the Deployment (the `kubectl` tool creates a
+{{<glossary_tooltip text="Service" term_id="service">}}  for you):
+```shell
+kubectl expose deployment configmap-two-containers --name=configmap-service --port=8080 --target-port=80
+```
+
+Use ```kubectl``` to forward the port:
+```shell
+kubectl port-forward service/configmap-service 8080:8080 & # this stays running in the background
+```
+
+Access the service.
+```shell
+curl http://localhost:8080
+```
+
`````
Is this Service creation relevant to the topic?

sftim replied:
It lets people see the impact of the ConfigMap change, right?

adityasamant25 replied:
Yes, the service creation is relevant. The section on multi-container pods is serving an html page on nginx. To enable http requests from outside the cluster we're creating a ClusterIP service and using the kubectl port-forward functionality to reach it. On minikube, I prefer using the port-forwarding than creating a NodePort service.

tengqm commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
```diff
@@ -0,0 +1,675 @@
+## Update configuration via a ConfigMap in a Pod possessing a sidecar container {#rollout-configmap-sidecar}
+
+The above scenario can be replicated by using a [Sidecar Container](/docs/concepts/workloads/pods/sidecar-containers/) as a helper container to write the HTML file.  
+As a Sidecar Container is conceptually an Init Container, it is guaranteed to start before the main webserver container.  
```
webserver is not an English word.

sftim replied:
(IMO, we can merge this PR even with the “webserver” word as is. I'm a native English speaker and I write “webserver” without the space fairly often).

adityasamant25 replied:
I will make the change from "webserver" to "web server" in the next commit.

tengqm commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,675 @@
+kubectl create configmap color --from-literal=color=blue
+```
+
+Below is an example manifest for a Deployment that manages a set of Pods, each with a main container and a Sidecar Container. The two containers share an `emptyDir` volume that they use to communicate.
````
Sidecar should be spelt as "sidecar" ?

adityasamant25 replied:
Ack. This will be handled in the next commit.

tengqm commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,675 @@
+
+Create a new ConfigMap by referencing the temporary manifest.
+```shell
+kubectl apply -f /var/folders/nn/bb6y1j8d69n9lhqppjlrfthm0000gn/T/kubectl-edit-1440698061.yaml
````
Engineering practice wise, I don't think we would encourage users to create a resource out of a temporary file. The result might be correct, but the practice should be discouraged.

sftim replied:
I agree; it's better to either:
- teach people to fetch the ConfigMap as YAML and edit it and apply the new manifest
- explain making a new immutable ConfigMap from a new manifest

I personally would also delete the old (immutable) ConfigMap once there's evidence that nothing is using it.

adityasamant25 replied:
Thanks for the advise. I have fixed this in the latest commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,675 @@
+kubectl expose deployment configmap-two-containers --name=configmap-service --port=8080 --target-port=80
+```
+
+Use ```kubectl``` to forward the port:
````
(nit)
```suggestion
Use `kubectl` to forward the port:
```

adityasamant25 replied:
Ack. Will be handled in the next commit.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,675 @@
+Loop over the service URL for few seconds. 
+```shell
+# Cancel this when you're happy with it (Ctrl-C)
+while true; do curl --connect-timeout 7.5 http://localhost:8080; sleep 10; done
````
Where do you run this? Does the `localhost` name resolve where people would expect?

adityasamant25 replied:
This is run from a local terminal (through which all the kubectl commands are executed). The `localhost` name resolves perfectly. I've tested it via minikube.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,675 @@
+
+Access the service.
+```shell
+curl http://localhost:8081
````
Where do you run this? Does the `localhost` name resolve where people would expect?

adityasamant25 replied:
This is run from a local terminal (through which all the kubectl commands are executed). The localhost name resolves perfectly. I've tested it via minikube.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,675 @@
+Loop over the service URL for few seconds.
+```shell
+# Cancel this when you're happy with it (Ctrl-C)
+while true; do curl --connect-timeout 7.5 http://localhost:8081; sleep 10; done
````
Where do you run this? Does the `localhost` name resolve where people would expect?

adityasamant25 replied:
This is run from a local terminal (through which all the kubectl commands are executed). The localhost name resolves perfectly. I've tested it via minikube.

sftim commented on content/en/docs/tutorials/configuration/updating-configuration-via-a-configmap.md:
````diff
@@ -0,0 +1,675 @@
+Sat Feb 17 13:13:35 UTC 2024 My preferred color is green
+```
+
+## Update configuration via an immutable ConfigMap that is mounted as a Volume {#rollout-configmap-immutable-volume}
````
nit: I wouldn't capitalize Volume as it is not an API kind.

adityasamant25 replied:
Ack. This will be handled in the next commit.

sftim commented:
@adityasamant25 this PR is nearly there for publication. Would you be willing to work on the last snags around the immutable ConfigMap part?

If you're short on time, you could drop “Fixes #23101” from the description, and leave out the part about immutable ConfigMaps from this PR. Once this one merges, you can open another PR about the immutable part and we can help you with just that bit.

I don't mind which way you go. Clearly, you've put in a lot of effort here. It's important that when we merge changes they're correct, which is why we're taking care over the updates to immutable ConfigMaps. We also don't like to close issues too early, which is why I pointed out what we shouldn't have merged the early version (it was marked as a full fix for #23101, but didn't yet cover immutable ConfigMap updates).

If you're on Kubernetes Slack and have questions about this, feel free to ask them (in #sig-docs). We really want to offer as much help as we have capacity for, as a SIG.

adityasamant25 commented:
> @adityasamant25 this PR is nearly there for publication. Would you be willing to work on the last snags around the immutable ConfigMap part?

Hi @sftim, Yes indeed. I'll work on applying the finishing touches here. 
If I fail to sort out the part about Immutable ConfigMaps fairly quickly, I'll extract that portion out of the PR.

adityasamant25 commented:
@sftim I have updated the immutable ConfigMaps section. I've also addressed the other review comments on this PR.

With these changes, I believe we are close to a merge.
I am available for any further modifications that may seem necessary.
I'm keen on having this page published soon.

adityasamant25 commented:
@nate-double-u As you are the PR wrangler for this week, may I please request you to review this PR.
In principle, it has been reviewed several times and all the review comments have been addressed.
I'm in need of a formal /lgtm post which @sftim will be able to approve it.

adityasamant25 commented:
@celestehorgan As you are the PR wrangler for this week, may I please request you to review this PR.
In principle, it has been reviewed several times and all the review comments have been addressed.
I'm in need of a formal /lgtm post which @sftim will be able to approve it.

dipesh-rawat reviewed (commented):
Nice work in creating this new tutorial page.

/lgtm

adityasamant25 commented:
Thank you @dipesh-rawat !

Hi @sftim, can you please approve the PR?

sftim commented:
/approve

sftim commented:
@adityasamant25, thank you for your effort and expertise on this PR.

adityasamant25 commented:
> @adityasamant25, thank you for your effort and expertise on this PR.

@sftim Thank you for the appreciation. Additionally, I want to genuinely thank you for your effort in reviewing the multiple iterations of this PR.
